### DISCUSSION TITLE
How we can use next safe action with generic types

### DISCUSSION THREAD

**Discussion by SpinBoxx** [next-safe-action/next-safe-action#176] [Q&A]
Helloo 🖐, 

i would like to use your lib with generic types. In my app, i have many datatables with button onCopy on each rows and i would like to create a server action where i pass the genereric return type in the function

Like this: 

copy.ts
```
"use server"
export const copyDataAction = createAction<RETURN_TYPE>
  .schema(dataSchema)
  .action(async ({ parsedInput: { dataId: string  } }): RETURN_TYPE => {
    const data = await fetch(`/api/${dataId}`)
    return data
  });
```

page.tsx

```
const { execute, result, status } = useAction<Note>(copyDataAction, {
   // Response take the generic return type 
    onSuccess: (response: Note) => {
      toast.success("ok");
    },
  });

// Return a note
 execute({
    dataId: "generic type string"
 });
```

I don't know if it yet possible, tell me 😇

Have a good day !

**Comment by TheEdoRan** [comment]
If I understood your request correctly, here's the solution. You can just pass the return type of the action to the `action` method when you're defining the function:

```typescript
type FetchResult = {
  response: object
}

export const testAction = ac.schema(z.object({
  username: z.string(),
  // pass generic return type here
})).action<FetchResult>(async () => {
  return {
    response: {
      yourData: "here"
    }
  }
});
```

**Reply by SpinBoxx** [reply]
Mmmmm not really, i need to pass the return type on the client and not in my server action 😅.
My goal is to pass it when i use my useAction hook
